assembly/src/parsers/io_ops.rs

```diff
@@ -256,69 +273,131 @@ pub fn parse_loc_storew(op: &Token) -> Result<Node, ParsingError> {
+                ));
+            }
+            Ok(value)
+        }
+        Err(_) => Err(ParsingError::invalid_param(op, param_idx)),
+    }
+}
+
+/// Parses a decimal parameter value into a u64 value valid for field element creation.
+fn parse_decimal_param(op: &Token, param_idx: usize) -> Result<u64, ParsingError> {
+    let value = parse_param::<u64>(op, param_idx)?;
+    if value >= Felt::MODULUS {
+        return Err(ParsingError::invalid_param_with_reason(
+            op,
+            param_idx,
+            format!("parameter value must be smaller than {}", Felt::MODULUS).as_str(),
+        ));
+    }
+    Ok(value)
+}
+
+/// Determines the minimal type appropriate for provided value and returns appropriate instruction
+/// for this value
+fn build_push_one_instruction(value: u64) -> Result<Node, ParsingError> {
+    if value <= u8::MAX as u64 {
+        Ok(Instruction(PushU8(value as u8)))
+    } else if value <= u16::MAX as u64 {
+        Ok(Instruction(PushU16(value as u16)))
+    } else if value <= u32::MAX as u64 {
+        Ok(Instruction(PushU32(value as u32)))
+    } else if value < Felt::MODULUS {
+        Ok(Instruction(PushFelt(Felt::new(value))))
     } else {
-        // iterate over the multi-value hex string and parse each 8-byte chunk into a valid element
-        (0..param_str.len())
-            .step_by(HEX_CHUNK_SIZE)
-            .map(|i| parse_hex_param(op, param_idx, &param_str[i..i + HEX_CHUNK_SIZE]))
-            .collect()
+        unreachable!()
+    }
+}
+
+/// Determines the minimal type appropriate for provided values and returns appropriate instruction
+/// for this values
+fn build_push_many_instruction(values: Vec<u64>) -> Result<Node, ParsingError> {
+    let values_max = values.iter().max();
+    match values_max {
+        Some(value) if *value <= u8::MAX as u64 => {
+            let values_u8 = values.iter().map(|v| *v as u8).collect();
+            Ok(Instruction(PushU8List(values_u8)))
+        }
+        Some(value) if *value <= u16::MAX as u64 => {
+            let values_u16 = values.iter().map(|v| *v as u16).collect();
+            Ok(Instruction(PushU16List(values_u16)))
+        }
+        Some(value) if *value <= u32::MAX as u64 => {
+            let values_u32 = values.iter().map(|v| *v as u32).collect();
+            Ok(Instruction(PushU32List(values_u32)))
+        }
+        Some(value) if *value < Felt::MODULUS => {
+            let mut values_felt = Vec::new();
+            values
+                .iter()
+                .for_each(|imm| values_felt.push(Felt::new(*imm)));
+            if values.len() == WORD_LEN {
+                Ok(Instruction(PushWord(
+                    values_felt.try_into().expect("Invalid constatnts length"),
+                )))
+            } else {
+                Ok(Instruction(PushFeltList(values_felt)))
+            }
+        }
+        _ => unreachable!(),
```

Comment:
If this is indeed unreachable, we should probably add an assert in the beginning of the function that `!values.is_empty()`.